compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -12216,84 +12213,174 @@ private boolean isMatchPatternEnd(SyntaxKind tokenKind) {
+            return STNodeFactory.createNodeList(argListMatchPatterns);
+        }
+        startContext(ParserRuleContext.ERROR_ARG_LIST_MATCH_PATTERN_FIRST_ARG);
+        STNode firstArg = parseErrorArgListMatchPattern(ParserRuleContext.ERROR_ARG_LIST_MATCH_PATTERN_START);
+        endContext();
+        if (isSimpleMatchPattern(firstArg.kind)) {
+
+            argListMatchPatterns.add(firstArg);
+            STNode argEnd = parseErrorArgListMatchPatternEnd(ParserRuleContext.ERROR_MESSAGE_MATCH_PATTERN_END);
+            if (argEnd != null) {
+                // null marks the end of args
+                STNode secondArg = parseErrorArgListMatchPattern(ParserRuleContext.ERROR_MESSAGE_MATCH_PATTERN_RHS);
+                if (isValidSecondArgMatchPattern(secondArg.kind)) {
+                    argListMatchPatterns.add(argEnd);
+                    argListMatchPatterns.add(secondArg);
+                } else {
+                    updateLastNodeInListWithInvalidNode(argListMatchPatterns, argEnd, null);
+                    updateLastNodeInListWithInvalidNode(argListMatchPatterns, secondArg,
+                            DiagnosticErrorCode.ERROR_MATCH_PATTERN_NOT_ALLOWED);
+                }
+            }
+        } else {
+            if (firstArg.kind != SyntaxKind.NAMED_ARG_MATCH_PATTERN &&
+                    firstArg.kind != SyntaxKind.REST_MATCH_PATTERN) {
+                addInvalidNodeToNextToken(firstArg, DiagnosticErrorCode.ERROR_MATCH_PATTERN_NOT_ALLOWED);
             } else {
-                updateLastNodeInListWithInvalidNode(argListMatchPatterns, currentArg, errorCode);
+                argListMatchPatterns.add(firstArg);
             }
+        }
 
-            STNode argRhs = parseArgMatchPatternRhs();
+        parseErrorFieldMatchPatterns(argListMatchPatterns);
+        return STNodeFactory.createNodeList(argListMatchPatterns);
+    }
+
+    private boolean isSimpleMatchPattern(SyntaxKind matchPatternKind) {
+        switch (matchPatternKind) {
+            case IDENTIFIER_TOKEN:
+            case SIMPLE_NAME_REFERENCE:
+            case NUMERIC_LITERAL:
+            case STRING_LITERAL:
+            case NULL_LITERAL:
+            case NIL_LITERAL:
+            case BOOLEAN_LITERAL:
+            case TYPED_BINDING_PATTERN:
+            case UNARY_EXPRESSION:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private boolean isValidSecondArgMatchPattern(SyntaxKind syntaxKind) {
+        switch (syntaxKind) {
+            case ERROR_MATCH_PATTERN:
+            case NAMED_ARG_MATCH_PATTERN:
+            case REST_MATCH_PATTERN:
+                return true;
+            default:
+                if (isSimpleMatchPattern(syntaxKind)) {
+                    return true;
+                }
+                return false;
+        }
+    }
 
-            if (argRhs == null) {
+    private void parseErrorFieldMatchPatterns(List<STNode> argListMatchPatterns) {
```

Comment:
It would be good to add the spec-snippet here as well, as a doc comment.
That would make it easier to quickly refer the grammar for someone who checks the code..

Reply:
added